akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala

```diff
@@ -452,6 +453,66 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
         toNet.expectNoBytes(100.millis) // the whole stage failed with bug #2236
       }
 
+      "handle WINDOW_UPDATE correctly when received before started sending out response and while receiving request data" in new WaitingForRequestData {
+        val bytesToSend = 70000 // > Http2Protocol.InitialWindowSize
+        val missingWindow = bytesToSend - Http2Protocol.InitialWindowSize
+        require(missingWindow >= 0)
+        sendWINDOW_UPDATE(0, missingWindow)
+        sendWINDOW_UPDATE(TheStreamId, missingWindow)
+
+        val entityDataOut = TestPublisher.probe[ByteString]()
+
+        val response = HttpResponse(entity = HttpEntity(ContentTypes.`application/octet-stream`, Source.fromPublisher(entityDataOut)))
+        emitResponse(TheStreamId, response)
+        expectDecodedResponseHEADERS(streamId = TheStreamId, endStream = false) shouldBe response.withEntity(HttpEntity.Empty.withContentType(ContentTypes.`application/octet-stream`))
+        entityDataOut.sendNext(bytes(bytesToSend, 0x23))
```

Comment:
How does this work? 
First, the server emits a response, and then it sends `bytesToSend` bytes? Doesn't that exceed the window size of `bytesToSend` octets?

Reply:
The user level interface uses a simple `Source[ByteString]` which can send any amount of data (I talked about that in our call this morning). The internal logic for outgoing streams is this:

 * frames are pulled one-by-one by the lower layers from the Http2ServerDemux
 * we keep a buffer of control frames to send (control frames are not subject to flow-control / windows) and a list of streams that can send data
 * if the multiplexer is pulled, first the control frames are drained and only if there are no pending control frames, one ready stream is chosen to send the next data frame
 * for each stream we have an outgoing buffer and if there's pending data the stream is listed to be chosen to send a frame
 * into that buffer, we pull from the user until we have at least `2 * currentMaxFrameSize` number of bytes buffered

Reply:
So, in case that wasn't clear, here we use the user-level interface to send a ByteString bigger than the WINDOW_SIZE and expect the rest to be buffered and only flushed later when more WINDOW is available.

Reply:
`entityDataOut` is a probe stream that is attached to the body of the `HttpResponse`. Only DATA frames count to the window size (https://tools.ietf.org/html/rfc7540#section-5.2.1 point 5), so this fits exactly